validator/src/journal/block_manager.rs

```diff
@@ -0,0 +1,1039 @@
+                    return Err(BlockManagerError::MissingPredecessor(format!(
+                        "During Put, missing predecessor of block {}: {}",
+                        head.header_signature, head.previous_block_id
+                    )));
+                } else {
+                    let mut previous = None;
+
+                    for block in tail {
+                        match previous {
+                            Some(previous_block_id) => {
+                                if block.previous_block_id != previous_block_id {
+                                    return Err(BlockManagerError::MissingPredecessorInBranch(
+                                        format!(
+                                            "During Put, missing predecessor of block {}: {}",
+                                            block.header_signature, block.previous_block_id
+                                        ),
+                                    ));
+                                }
+                                previous = Some(block.header_signature.clone());
+                            }
+                            None => {
+                                if block.previous_block_id != head.header_signature {
+                                    return Err(BlockManagerError::MissingPredecessorInBranch(
+                                        format!(
+                                            "During Put, missing predecessor of block {}: {}",
+                                            block.previous_block_id, head.header_signature
+                                        ),
+                                    ));
+                                }
+
+                                previous = Some(block.header_signature.clone());
+                            }
+                        }
+                    }
+
+                    match self.block_by_block_id
+                        .get_mut(head.previous_block_id.as_str())
+                    {
+                        Some(mut tup) => {
+                            let &mut (_, ref mut internal_ref_count, _) = tup;
+                            *internal_ref_count += 1;
+                        }
+                        None => {
+                            self.anchors
+                                .increase_internal_ref_count(head.previous_block_id.as_str());
+                        }
+                    }
+                }
+            }
+            None => return Err(BlockManagerError::MissingInput),
+        }
+        let blocks_not_added_yet: Vec<Block> = branch
+            .into_iter()
+            .filter(|block| {
+                !self.block_by_block_id.contains_key(&block.header_signature)
+                    && !self.blockstore_by_name.values().any(|blockstore| {
+                        blockstore.get(vec![block.header_signature.clone()]).count() > 0
+                    })
+            })
+            .collect();
```

Comment:
Can you avoid the extra vector allocation here by chaining a for_each() here instead? The you can just do `if let Some(last_block) = branch.last()` after?